### PR TITLE
Update icpr_cert for requesting certs on behalf of users

### PR DESCRIPTION
**Pull request by zeroSteiner**
This updates the existing `icpr_cert` module, building on the work started in #16939. This allows it to request certificates on behalf of other users which is notably necessary for exploiting [ESC2 and ESC3](https://posts.specterops.io/certified-pre-owned-d95910965cd2).

This requires the changes from rapid7/ruby_smb#239 to fix a bug that occurs while issuing the certificate.

I'm debating whether additional data definitions defined in `Rex::Proto::Kerberos::Model::Pkinit` should be moved into the new `Rex::Proto::CryptoAsn1` module since they're being used for things other than Kerberos PKINIT. Those definitions would be `SignedData`, `ContentInfo`, `SignerInfo` and their member definitions.

## Verification

- [ ] Install ADCS on either a new or existing domain controller (steps copied from #16939)
    - [ ] Open the Server Manager
    - [ ] Select Add roles and features
    - [ ] Select "Active Directory Certificate Services" under the "Server Roles" section
    - [ ] When prompted add all of the features and management tools
    - [ ] On the AD CS "Role Services" tab, leave the default selection of only "Certificate Authority"
    - [ ] Completion the installation and reboot the server
    - [ ] Reopen the Server Manager
    - [ ] Go to the AD CS tab and where it says "Configuration Required", hit "More" then "Configure Active Directory Certificate..."
    - [ ] Select "Certificate Authority" in the Role Services tab
    - [ ] Keep all of the default settings, noting the value of the "Common name for this CA" on the "CA Name" tab (this value corresponds to the `CA` datastore option)
    - [ ] Accept the rest of the default settings and complete the configuration
- [ ] Configure a certificate template to be vulnerable via ESC2, use the steps from this excellent reference: [RayRRT/Active-Directory-Certificate-Services-abuse](https://github.com/RayRRT/Active-Directory-Certificate-Services-abuse#heading5)
- [ ] Configure a certificate template to be vulnerable via ESC3, use the steps from this excellent reference: [RayRRT/Active-Directory-Certificate-Services-abuse](https://github.com/RayRRT/Active-Directory-Certificate-Services-abuse#heading6)
- [ ] Use the module and follow the steps from the documentation, the module must be run twice to exploit ESC2 or ESC3. The steps are the same in both cases.
    - [ ] Set the `RHOSTS`, `CA`, `SMBUser` and `SMBPass` options
    - [ ] Set `CERT_TEMPLATE` to the vulnerable template (e.g. `ESC2-Test` or `ESC3-Test`)
    - [ ] Run the module and receive a certificate
    - [ ] Set the `PFX` file to the issued certificate
    - [ ] Set the `ON_BEHALF_OF` option to a user to target (typically a domain admin). The format is `DOMAIN\USERNAME`
    - [ ] Set the `CERT_TEMPLATE` option to another template, the default `User` one should work
    - [ ] Run the module and receive a second certificate for the target user
    - [ ] Test the second certificate allows you to authenticate as the target user using the new `pkinit_login` module (see #17117 for details)

### CODE REVIEW

File: modules/auxiliary/admin/dcerpc/icpr_cert.rb [hunk 1]
```diff
@@ -195,10 +234,81 @@ def build_csr(cn:, private_key:, dns: nil, msext_upn: nil)
...
 
+  # Make a certificate request on behalf of another user.
+  #
+  # @param [OpenSSL::X509::Request] csr The certificate request to make on behalf of the user.
+  # @param [String] on_behalf_of The user to make the request on behalf of.
+  # @param [OpenSSL::X509::Certificate] cert The public key to use for signing the request.
+  # @param [OpenSSL::PKey::RSA] key The private key to use for signing the request.
+  # @param [String] algorithm The digest algorithm to use.
+  # @return [Rex::Proto::Kerberos::Model::Pkinit::ContentInfo] The signed request content.
+  def build_on_behalf_of(csr:, on_behalf_of:, cert:, key:, algorithm: 'SHA256')
```
Comment (adfoster-r7) [inline modules/auxiliary/admin/dcerpc/icpr_cert.rb:249]: It'd be great to add a quick spec for this method to verify that the `ContentInfo` der is as you expect it - just to catch any shifts in the upstream library

File: lib/rex/proto/crypto_asn1.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    def value_to_der
+      @value.to_s.dup.encode('UTF-16BE').b
+    end
+
+    def der_to_value(der, ber: false)
+      super
+      @value = der.dup.force_encoding('UTF-16BE')
+    end
+  end
+
```
Comment (adfoster-r7) [inline lib/rex/proto/crypto_asn1.rb:41]: A docs link to the spec would be great to add here for future folk that aren't familiar with what model this is implementing 🎉

File: modules/auxiliary/admin/dcerpc/icpr_cert.rb (hunk 1 above)
Comment (adfoster-r7) [reply modules/auxiliary/admin/dcerpc/icpr_cert.rb:249]: Or if it's easier to just integration test the full module, and assert that `store_loot` is called with the right der bytes - that works for me too :+1:

Comment (zeroSteiner) [reply modules/auxiliary/admin/dcerpc/icpr_cert.rb:249]: Added in 749bfa2aadbd96ca33800d53da34dec640b1a2c2, let me know if that's not what you had in mind.

File: modules/auxiliary/admin/dcerpc/icpr_cert.rb [hunk 3]
```diff
@@ -120,15 +136,37 @@ def run
     fail_with(Failure::Unknown, e.message)
   end
 
+  def setup_request_cert
+    if datastore['ON_BEHALF_OF'].present?
+      errors = {}
+      errors['ON_BEHALF_OF'] = 'Must be in the format DOMAIN\\USER.' unless datastore['ON_BEHALF_OF'] =~ /\S+\\\S+/
```
Comment (smcintyre-r7) [inline modules/auxiliary/admin/dcerpc/icpr_cert.rb:142]: Let's make this check more restrictive.

File: modules/auxiliary/admin/dcerpc/icpr_cert.rb [hunk 4]
```diff
@@ -50,8 +55,13 @@ def initialize(info = {})
       OptString.new('CERT_TEMPLATE', [ true, 'The certificate template', 'User' ]),
       OptString.new('ALT_DNS', [ false, 'Alternative certificate DNS' ]),
       OptString.new('ALT_UPN', [ false, 'Alternative certificate UPN' ]),
```
Comment (smcintyre-r7) [inline modules/auxiliary/admin/dcerpc/icpr_cert.rb:57]:
```suggestion
      OptString.new('ALT_UPN', [ false, 'Alternative certificate UPN (format: USER@DOMAIN)' ]),
```

**Review by adfoster-r7 — APPROVED** [review]
Looks good to me from a code review perspective - but I won't merge this or assign it to myself as I haven't had a cycle to test this before I jump out for a week. Happy to pass on the reins to anyone wants to test the PR and land it in the mean time:+1:

**Review by cdelafuente-r7 — COMMENTED** [review]
Thanks @zeroSteiner for updating this module. I just left a few minor comments before it lands. I tested the new ESC2 and ESC3 attack scenarios and it works great.

File: modules/auxiliary/admin/dcerpc/icpr_cert.rb [hunk 5]
```diff
@@ -195,10 +239,81 @@ def build_csr(cn:, private_key:, dns: nil, msext_upn: nil)
...
+      sid: {
+        issuer: cert.issuer,
+        serial_number: cert.serial.to_i
+      },
+      digest_algorithm: {
+        algorithm: digest_oid
+      },
+      signed_attrs: [
+        {
+          attribute_type: '1.3.6.1.4.1.311.13.2.1',
```
[the remark this answers is not in the document]
Comment (zeroSteiner) [reply modules/auxiliary/admin/dcerpc/icpr_cert.rb:274]: Defined the constant at the top of the module with the other module-specific OIDs so it's more readable.

**Review by cdelafuente-r7 — APPROVED** [review]

**Comment by cdelafuente-r7** [comment]
# Release Notes
This updates the existing `icpr_cert` module to request certificates on behalf of other users which is notably necessary for exploiting ESC2 and ESC3.

**Pull request merged**